Taco Notebook.ipynb

```diff
@@ -0,0 +1,587 @@
+   "source": [
+    "# Defining Tensor Formats <a class=\"anchor\" id=\"second-section\"></a>\n",
+    "\n",
+    "Consider a matrix $M$ (aka a two-dimensional tensor) containing the following values:\n",
+    "\n",
+    "$$M = \\begin{bmatrix}\n",
+    "    6 & \\cdot & 9 & 8 \\\\\n",
+    "    \\cdot & \\cdot & \\cdot & \\cdot \\\\\n",
+    "    5 & \\cdot & \\cdot & 7  \n",
+    "\\end{bmatrix}$$\n",
+    "\n",
+    "Denote the rows and columns as dimensions $d_1$ and $d_2$, respectively. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We look at how $M$ is represented differently in different formats. For convenience, let's define a helper function to initialize $M$. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def make_example_matrix(format):\n",
+    "    M = pt.tensor([3, 4], format)\n",
+    "    M.insert([0, 0], 6)\n",
+    "    M.insert([0, 2], 9)\n",
+    "    M.insert([0, 3], 8)\n",
+    "    M.insert([2, 0], 5)\n",
+    "    M.insert([2, 3], 7)\n",
+    "    return M"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## (dense $d_1$, dense $d_2$)\n",
+    "Note that passing in `dense` makes all of the dimensions dense. This is equivalent to `pt.format(dense, dense)`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "make_example_matrix(dense)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For this example, we focus on the last line of the output, the `vals` array: since all values are stored, it is simply a flattened $3 \\times 4$ matrix."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## (dense $d_1$, compressed $d_2$)\n",
+    "This is called compressed sparse row (CSR) format."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "csr = pt.format([dense, compressed])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "make_example_matrix(csr)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Since $d_1$ is dense, we need only store the `size` of the dimension; values, both zero and nonzero, are stored for every coordinate in dimension $d_1$.  \n",
+    "\n",
+    "Since $d_2$ is compressed, we store a `pos` array and an `idx` array; these together form a segmented vector with one segment per entry in the previous dimension. The `idx` array stores all the indices with nonzero values in the dimension, while the `pos` array stores the location in the `idx` array where each segment begins. In particular, segment $i$ is stored in locations `pos[i]:pos[i+1]` in the `idx` array.\n",
```

Comment:
"store a `pos` array and an `idx` array" -> "store a `pos` array (`[0, 3, 3, 5]`) and an `idx` array (`[0, 2, 3, 0, 3]`)"